src/rt.rs

```diff
@@ -1,43 +1,124 @@
 use std::mem;
 
+use super::Result;
 use futures::Future;
+use std::pin::Pin;
 use tokio::runtime::Runtime;
+use tokio::task::JoinHandle;
 use tokio::task::LocalSet;
 use tokio::time::{sleep, Duration, Instant};
 
-/// Per host simulated runtime
+// To support re-creation, we need to store a factory of the future that
+// represents the software. This is somewhat annoying in that it requires
+// boxxing to avoid generics.
+type Software<'a> = Box<dyn Fn() -> Pin<Box<dyn Future<Output = Result>>> + 'a>;
+
+/// Type of runtime
+enum Kind<'a> {
+    /// A runtime that is running client software
+    Client,
+
+    /// A runtime that is running simulated host software
+    Host { software: Software<'a> },
+
+    /// A runtime that with no software running
+    NoSoftware,
+}
+
+/// Per host simulated runtime that runs software
 ///
 /// The tokio runtime is paused (see [`Builder::start_paused`]), which gives us
 /// control over when and how to advance time. In particular, see [`Rt::tick`],
 /// which lets the runtime do a bit more work.
-pub(crate) struct Rt {
+pub(crate) struct Rt<'a> {
+    kind: Kind<'a>,
+
     /// Handle to the Tokio runtime driving this simulated host. Each runtime
     /// may have a different sense of "now" which simulates clock skew.
     tokio: Runtime,
 
     /// Local task set used for running !Send tasks.
     local: LocalSet,
+
+    /// Software's handle
+    /// The handle is consumed as soon as software finishes or host crashes,
+    /// replaced with None to indicate finished software.
+    handle: Option<JoinHandle<Result>>,
 }
 
-impl Rt {
-    pub(crate) fn new() -> Self {
+impl<'a> Rt<'a> {
+    pub(crate) fn client<F>(client: F) -> Self
+    where
+        F: Future<Output = Result> + 'static,
+    {
+        let (tokio, local) = init();
+
+        let handle = with(&tokio, &local, || tokio::task::spawn_local(client));
+
+        Self {
+            kind: Kind::Client,
+            tokio,
+            local,
+            handle: Some(handle),
+        }
+    }
+
+    pub(crate) fn host<F, Fut>(software: F) -> Self
+    where
+        F: Fn() -> Fut + 'a,
+        Fut: Future<Output = Result> + 'static,
+    {
+        let (tokio, local) = init();
+
+        let software: Software = Box::new(move || Box::pin(software()));
+
+        let handle = with(&tokio, &local, || tokio::task::spawn_local(software()));
+
+        Self {
+            kind: Kind::Host { software },
+            tokio,
+            local,
+            handle: Some(handle),
+        }
+    }
+
+    // Creates special instance of `Rt` that has no running software and used
+    // only by Topology to advance simulated time.
```

Comment:
```suggestion
```

I'd suggest either documenting all of the constructors or none. The docs on `Kind` likely do the trick.